src/vma/dev/ring_bond.cpp

```diff
@@ -486,110 +479,42 @@ void ring_bond::devide_buffers_helper(mem_buf_desc_t *p_mem_buf_desc_list, mem_b
 	}
 }
 
-/* TODO consider only ring_simple to inherit mem_buf_desc_owner */
-void ring_bond::mem_buf_desc_completion_with_error_rx(mem_buf_desc_t* p_rx_wc_buf_desc)
+void ring_bond::popup_active_rings()
 {
-	NOT_IN_USE(p_rx_wc_buf_desc);
-	ring_logpanic("programming error, how did we got here?");
-}
+	ring_slave *cur_slave = NULL;
+	int i, j;
 
-void ring_bond::mem_buf_desc_completion_with_error_tx(mem_buf_desc_t* p_tx_wc_buf_desc)
-{
-	NOT_IN_USE(p_tx_wc_buf_desc);
-	ring_logpanic("programming error, how did we got here?");
-}
-
-void ring_bond::mem_buf_desc_return_to_owner_rx(mem_buf_desc_t* p_mem_buf_desc, void* pv_fd_ready_array /*NULL*/)
-{
-	NOT_IN_USE(p_mem_buf_desc);
-	NOT_IN_USE(pv_fd_ready_array);
-	ring_logpanic("programming error, how did we got here?");
-}
-
-void ring_bond::mem_buf_desc_return_to_owner_tx(mem_buf_desc_t* p_mem_buf_desc)
-{
-	NOT_IN_USE(p_mem_buf_desc);
-	ring_logpanic("programming error, how did we got here?");
-}
-
-void ring_bond_eth::create_slave_list(in_addr_t local_if, ring_resource_creation_info_t* p_ring_info, bool active_slaves[], uint16_t vlan)
-{
-	for (uint32_t i = 0; i < m_n_num_resources; i++) {
-		m_bond_rings[i] = new ring_eth(local_if, &p_ring_info[i], 1, active_slaves[i], vlan, m_mtu, this);
-		if (m_min_devices_tx_inline < 0)
-			m_min_devices_tx_inline = m_bond_rings[i]->get_max_tx_inline();
-		else
-			m_min_devices_tx_inline = min(m_min_devices_tx_inline, m_bond_rings[i]->get_max_tx_inline());
-		if (active_slaves[i]) {
-			m_active_rings[i] = m_bond_rings[i];
-		} else {
-			m_active_rings[i] = NULL;
-		}
-	}
-	close_gaps_active_rings();
-}
-
-void ring_bond_ib::create_slave_list(in_addr_t local_if, ring_resource_creation_info_t* p_ring_info, bool active_slaves[], uint16_t pkey)
-{
-	for (uint32_t i = 0; i < m_n_num_resources; i++) {
-		m_bond_rings[i] = new ring_ib(local_if, &p_ring_info[i], 1, active_slaves[i], pkey, m_mtu, this); // m_mtu is the value from ifconfig when ring created. Now passing it to its slaves. could have sent 0 here, as the MTU of the bond is already on the bond
-		if (m_min_devices_tx_inline < 0)
-			m_min_devices_tx_inline = m_bond_rings[i]->get_max_tx_inline();
-		else
-			m_min_devices_tx_inline = min(m_min_devices_tx_inline, m_bond_rings[i]->get_max_tx_inline());
-		if (active_slaves[i]) {
-			m_active_rings[i] = m_bond_rings[i];
-		} else {
-			m_active_rings[i] = NULL;
+	for (i = 0; i < (int)m_bond_rings.size(); i++) {
+		for (j = i + 1; j < (int)m_bond_rings.size(); j++) {
+			if (!m_bond_rings[i]->m_active && m_bond_rings[j]->m_active) {
+				cur_slave = m_bond_rings[i];
+				m_bond_rings[i] = m_bond_rings[j];
+				m_bond_rings[j] = cur_slave;
+			}
```

Comment:
Complexity can be reduce to O(N) - instead of O(N^2) :
J loop should start from the end and not for i + 1